README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# colortable <img src='man/figures/logo.png' align="right" height="200" />

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![R build
status](https://github.com/thebioengineer/colortable/workflows/R-CMD-check/badge.svg)](https://github.com/thebioengineer/colortable/actions)
[![Codecov test
coverage](https://codecov.io/gh/thebioengineer/colortable/branch/master/graph/badge.svg)](https://codecov.io/gh/thebioengineer/colortable?branch=master)
<!-- badges: end -->

Seamlessly style and print your vectors across Rmarkdown output types
through a single interface. colortable enables users to style and color
the contents of their vectors, data.frames, and tibbles through the
object, and function, `color_vctr()`.

At this point, the supported output types include:

<div style="display:inline-block">

<img alt="console" src="inst/media/console_logo.png" height = 80/>
<img alt="html" src="inst/media/html_logo.png" height = 80/>
<img alt="pdf" src="inst/media/pdf_logo.png" height = 80/>
<img alt="Microsoft Word" src="inst/media/microsoft_word_logo.png" height = 80/>
<img alt="Beamer Presentations" src="inst/media/beamer_pres.png" height = 80/>

</div>

## Installation

<!-- You can install the released version of colortable from [CRAN](https://CRAN.R-project.org) with: -->

Currently {colortable} is only available on github, and is very much
under development.

``` r
remotes::install_github("thebioengineer/colortable")
## install.packages("colortable") ## Not Available on CRAN
```

## Getting Started

{colortable} works by making a special S3 class called a `color_vctr`,
and custom print/format functions.

It then has 4 arguments; - The vector to be styled - text\_color, a
vector that is either the color the entire vector to be colored or each
element - style, a vector is either the style the enture vector to be
styled with or each element - background, a vector that is either the
background color the entire vector to be colored or for each element.

Additionally, there is a few helper functions - `set_styling()` uses a
boolean argument to apply the styling - `color_scale()` is to be used
for setting colors, accepting a pallette

*A note, html styling does not apply in a github readme*

``` r
library(colortable)
#> Registered S3 methods overwritten by 'colortable':
#>   method                from     
#>   knit_print.data.frame rmarkdown
#>   print.data.frame      base

color_vctr(c(1,2,3,4),
           text_color = c("blue","green", "yellow",NA),
           style = c("underline","italic",NA,"bold"),
           background = c(NA,NA,"blue",NA))
```

<pre><code class = "hljs"><span>## [1] <span style='text-decoration:underline;color:blue;'>1</span> <span style='font-style:italic;color:lime;'>2</span> <span style='color:yellow;background:blue;'>3</span> <span style='font-weight:bold;'>4</span> </span><br></code></pre>

``` r

color_vctr(LETTERS, text_color = color_scale(colorRamp(c("red","yellow"))))
```

<pre><code class = "hljs"><span>## [1] <span style='color:#FF0000;'>A</span> <span style='color:#FF0A00;'>B</span> <span style='color:#FF1400;'>C</span> <span style='color:#FF1F00;'>D</span> <span style='color:#FF2900;'>E</span> <span style='color:#FF3300;'>F</span> <span style='color:#FF3D00;'>G</span> <span style='color:#FF4700;'>H</span> <span style='color:#FF5200;'>I</span> <span style='color:#FF5C00;'>J</span> <span style='color:#FF6600;'>K</span> <span style='color:#FF7000;'>L</span> <span style='color:#FF7A00;'>M</span> <span style='color:#FF8500;'>N</span> <span style='color:#FF8F00;'>O</span> <span style='color:#FF9900;'>P</span> <span style='color:#FFA300;'>Q</span> <span style='color:#FFAD00;'>R</span> <span style='color:#FFB800;'>S</span> <span style='color:#FFC200;'>T</span> <span style='color:#FFCC00;'>U</span> <span style='color:#FFD600;'>V</span> <span style='color:#FFE000;'>W</span> <span style='color:#FFEB00;'>X</span> <span style='color:#FFF500;'>Y</span> <span style='color:#FFFF00;'>Z</span> </span><br></code></pre>

``` r
color_vctr(LETTERS, text_color = color_scale("Blues"))
```

<pre><code class = "hljs"><span>## [1] <span style='color:#F7FBFF;'>A</span> <span style='color:#EFF6FC;'>B</span> <span style='color:#E7F1FA;'>C</span> <span style='color:#DFECF7;'>D</span> <span style='color:#D7E6F5;'>E</span> <span style='color:#D0E1F2;'>F</span> <span style='color:#C8DCF0;'>G</span> <span style='color:#BDD7EC;'>H</span> <span style='color:#B0D1E7;'>I</span> <span style='color:#A3CCE3;'>J</span> <span style='color:#94C4DF;'>K</span> <span style='color:#85BBDB;'>L</span> <span style='color:#74B2D8;'>M</span> <span style='color:#65A9D3;'>N</span> <span style='color:#59A0CE;'>O</span> <span style='color:#4B98C9;'>P</span> <span style='color:#3F8EC4;'>Q</span> <span style='color:#3583BF;'>R</span> <span style='color:#2B79B9;'>S</span> <span style='color:#206EB3;'>T</span> <span style='color:#1964AB;'>U</span> <span style='color:#115AA3;'>V</span> <span style='color:#08509A;'>W</span> <span style='color:#09458A;'>X</span> <span style='color:#093A7A;'>Y</span> <span style='color:#08306B;'>Z</span> </span><br></code></pre>

## Example

A common case I have seen for coloring values is from analysis coloring
p-values. Normally, when I have seen this the color is hard-coded in an
ifelse statement with a paste0. However, this liits the output to a
single type.

The benefit of {{colortable}} is that the same code can be used across
outputs and even in the console\!

``` r
library(tidyverse)
library(knitr)

## Super Great analysis of mtcars!

lm_fit <- lm(mpg ~ ., mtcars)

a_lm_fit <- anova(lm_fit)

tbl_anova <- a_lm_fit %>% 
  as_tibble()%>% 
  mutate(
    Coef     = rownames(a_lm_fit),
    `Pr(>F)` = set_styling(`Pr(>F)`, `Pr(>F)` < 0.05, background = "green", style = "underline"),
    `Pr(>F)` = set_styling(`Pr(>F)`, is.na(`Pr(>F)`), style = "strikethrough", text_color = "silver"),
    `F value` = set_styling(`F value`, is.na(`F value`), style = "strikethrough", text_color = "silver")
  ) %>% 
  select(Coef, everything())

kable(tbl_anova, escape = FALSE)
```

![examples](inst/media/output_gifs.gif)

## Output types

In order to simply generate a color\_vctr, use the `color_vctr`
function. It can convert any atomic (numeric, integer, complex,
character, logical, raw) into a color\_vctr where text and background
colors, and styles can be set.

To see the available styles and colors, use the `valid_*` family of
functions: `valid_colors()`or `valid_style()`. To check whether the
styling is a valid type for the output, set the method to be “latex” for
pdf outputs, or “html” for html outputs.

Below is a random sampling of output types to the console:

``` r

data.frame(
  text_color = sample(c(NA, valid_text_color()),10, replace = TRUE),
  background = sample(c(NA, valid_background()),10, replace = TRUE),
  style      = sample(c(NA, valid_style()),10, replace = TRUE),
  stringsAsFactors = FALSE
  ) %>% 
  mutate(
    background = ifelse(text_color == background, 
                  sample(c(NA, valid_background()),10, replace = TRUE),
                  background)
  ) %>% 
  mutate(
  example = color_vctr(runif(10),
                       text_color = text_color,
                       background = background,
                       style = style)
  )
```

![examples](inst/media/multiple_output_types.gif)

## Inspiration

This idea was inspired by [`crayon`](https://github.com/r-lib/crayon),
and has some elements based on it. I thank all the developers of that
project\! Since then, I have been insprired by
[‘flextable’](https://github.com/davidgohel/officedown) for word
development.

Current styling technologies such as {kableExtra} and {formattable} also
inspired the development of this project.

## COC

Please note that the ‘colortable’ project is released with a
[Contributor Code of Conduct](CODE_OF_CONDUCT.md). By contributing to
this project, you agree to abide by its terms.


NEWS.md
# colortable 0.3.0

### Features
  - Initial implementation of word (docx) printing
  - knit_print abilities to color_vctrs
  
# colortable 0.2.1

### Features
  - Improved vector printing to console
  - Improved data.frame printing to console

### Testing
  - Additional printing tests added

# colortable 0.2.0

### Features
  * Print to many more types of outputs
      - 255 console color types
      - 148 named colors in HTML, but any hexcode
      - 718 colors in pdf
  
  * Add `color_scale()` that can be used to set a color scaling of a color_vctr
  
  * Improved printing in notebooks for inline **still need to use kable to print

# colortable 0.1.1

# colortable 0.1.0

* Initial commit of the {{colortable}} package to github. Many things are under development!
* Added a `NEWS.md` file to track changes to the package.


R/style_html.R


style_wrapper_html <- function(styling, type = c("text","style","background")){
  if (is.na(styling)){
    ""
  }else{
    type <- match.arg(type)
    styling <- tolower(styling)
    switch(
      type,
      "text" = html_text_styling(unify_colors(styling, type = "html")),
      "style" = html_decoration_styling(styling),
      "background" = html_background_styling(unify_colors(styling, type = "html"))
    )
  }
}

html_decoration_styling <- function(styling){
  if (!styling %in% names(html_style_codes)) {
    ""
  }else{
    codes <- html_style_codes[[styling]]
    paste0(names(codes),":",codes,";",collapse = "")
  }
}

html_text_styling <- function(color){
  codes <- try(find_html_codes(color),silent = TRUE)
  if (inherits(codes, "try-error")) {
    ""
  }else{
    paste0("color:",codes,";",collapse = "")
  }
}

html_background_styling <- function(color){
  codes <- try(find_html_codes(color),silent = TRUE)
  if (inherits(codes, "try-error")) {
    ""
  } else{
    paste0("background:",codes,";",collapse = "")
  }
}

valid_html_codes <- function(color) {
  valid <- FALSE
  if (grepl("^#", color) & grepl("^#[0-9A-F]{6}$", color, perl = TRUE)) {
    valid <- TRUE
  } else if (color %in% color_key_html$Name) {
    valid <- TRUE
  } else{
    valid <- FALSE
  }
  return(valid)
}


find_html_codes <- function(input){
  if (valid_html_codes(input)) {
    return(input)
  }else{
    u_inputs <- unique(input)
    output_codes <- vector("character", length(input))
    rgb_mat <- col2rgb(u_inputs)
    rgb_key <- do.call(rbind, color_key_html$RGB)
    for (input_i in 1:length(u_inputs)) {
      input_rgb <- rgb_mat[, input_i, drop = TRUE]
      output_codes[input == u_inputs[input_i]] <-
        color_key_html$Name[which_closest_color(input_rgb, rgb_key)]
    }

    output_codes
  }
}

html_style_codes <- list(
  bold = list("font-weight" = "bold"),
  italic = list("font-style" = "italic"),
  underline = list("text-decoration" = "underline"),
  inverse = list("-webkit-filter" = "invert(100%)",
                 "filter" = "invert(100%)"),
  hidden = list("visibility" = "hidden"),
  strikethrough = list("text-decoration" = "line-through"),
  outline = list("-webkit-text-stroke" = "1px black",
                       "text-shadow" = "0px 0px 0px #000")
)

#' return the tibble of valid web colors
#'
#' Source is from https://www.w3schools.com/colors/color_tryit.asp
#'
#' @importFrom tibble tribble
#'
color_key_html  <- tibble::tribble(
              ~Name,     ~hex,          ~RGB,
             "aliceblue", "#F0F8FF", c(red = 240, green = 248, blue = 255),
          "antiquewhite", "#FAEBD7", c(red = 250, green = 235, blue = 215),
                  "aqua", "#00FFFF",   c(red = 0, green = 255, blue = 255),
            "aquamarine", "#7FFFD4", c(red = 127, green = 255, blue = 212),
                 "azure", "#F0FFFF", c(red = 240, green = 255, blue = 255),
                 "beige", "#F5F5DC", c(red = 245, green = 245, blue = 220),
                "bisque", "#FFE4C4", c(red = 255, green = 228, blue = 196),
                 "black", "#000000",       c(red = 0, green = 0, blue = 0),
        "blanchedalmond", "#FFEBCD", c(red = 255, green = 235, blue = 205),
                  "blue", "#0000FF",     c(red = 0, green = 0, blue = 255),
            "blueviolet", "#8A2BE2",  c(red = 138, green = 43, blue = 226),
                 "brown", "#A52A2A",   c(red = 165, green = 42, blue = 42),
             "burlywood", "#DEB887", c(red = 222, green = 184, blue = 135),
             "cadetblue", "#5F9EA0",  c(red = 95, green = 158, blue = 160),
            "chartreuse", "#7FFF00",   c(red = 127, green = 255, blue = 0),
             "chocolate", "#D2691E",  c(red = 210, green = 105, blue = 30),
                 "coral", "#FF7F50",  c(red = 255, green = 127, blue = 80),
        "cornflowerblue", "#6495ED", c(red = 100, green = 149, blue = 237),
              "cornsilk", "#FFF8DC", c(red = 255, green = 248, blue = 220),
               "crimson", "#DC143C",   c(red = 220, green = 20, blue = 60),
                  "cyan", "#00FFFF",   c(red = 0, green = 255, blue = 255),
              "darkblue", "#00008B",     c(red = 0, green = 0, blue = 139),
              "darkcyan", "#008B8B",   c(red = 0, green = 139, blue = 139),
         "darkgoldenrod", "#B8860B",  c(red = 184, green = 134, blue = 11),
              "darkgray", "#A9A9A9", c(red = 169, green = 169, blue = 169),
              "darkgrey", "#A9A9A9", c(red = 169, green = 169, blue = 169),
             "darkgreen", "#006400",     c(red = 0, green = 100, blue = 0),
             "darkkhaki", "#BDB76B", c(red = 189, green = 183, blue = 107),
           "darkmagenta", "#8B008B",   c(red = 139, green = 0, blue = 139),
        "darkolivegreen", "#556B2F",   c(red = 85, green = 107, blue = 47),
            "darkorange", "#FF8C00",   c(red = 255, green = 140, blue = 0),
            "darkorchid", "#9932CC",  c(red = 153, green = 50, blue = 204),
               "darkred", "#8B0000",     c(red = 139, green = 0, blue = 0),
            "darksalmon", "#E9967A", c(red = 233, green = 150, blue = 122),
          "darkseagreen", "#8FBC8F", c(red = 143, green = 188, blue = 143),
         "darkslateblue", "#483D8B",   c(red = 72, green = 61, blue = 139),
         "darkslategray", "#2F4F4F",    c(red = 47, green = 79, blue = 79),
         "darkslategrey", "#2F4F4F",    c(red = 47, green = 79, blue = 79),
         "darkturquoise", "#00CED1",   c(red = 0, green = 206, blue = 209),
            "darkviolet", "#9400D3",   c(red = 148, green = 0, blue = 211),
              "deeppink", "#FF1493",  c(red = 255, green = 20, blue = 147),
           "deepskyblue", "#00BFFF",   c(red = 0, green = 191, blue = 255),
               "dimgray", "#696969", c(red = 105, green = 105, blue = 105),
               "dimgrey", "#696969", c(red = 105, green = 105, blue = 105),
            "dodgerblue", "#1E90FF",  c(red = 30, green = 144, blue = 255),
             "firebrick", "#B22222",   c(red = 178, green = 34, blue = 34),
           "floralwhite", "#FFFAF0", c(red = 255, green = 250, blue = 240),
           "forestgreen", "#228B22",   c(red = 34, green = 139, blue = 34),
               "fuchsia", "#FF00FF",   c(red = 255, green = 0, blue = 255),
             "gainsboro", "#DCDCDC", c(red = 220, green = 220, blue = 220),
            "ghostwhite", "#F8F8FF", c(red = 248, green = 248, blue = 255),
                  "gold", "#FFD700",   c(red = 255, green = 215, blue = 0),
             "goldenrod", "#DAA520",  c(red = 218, green = 165, blue = 32),
                  "gray", "#808080", c(red = 128, green = 128, blue = 128),
                  "grey", "#808080", c(red = 128, green = 128, blue = 128),
                 "green", "#008000",     c(red = 0, green = 128, blue = 0),
           "greenyellow", "#ADFF2F",  c(red = 173, green = 255, blue = 47),
              "honeydew", "#F0FFF0", c(red = 240, green = 255, blue = 240),
               "hotpink", "#FF69B4", c(red = 255, green = 105, blue = 180),
             "indianred", "#CD5C5C",   c(red = 205, green = 92, blue = 92),
                "indigo", "#4B0082",    c(red = 75, green = 0, blue = 130),
                 "ivory", "#FFFFF0", c(red = 255, green = 255, blue = 240),
                 "khaki", "#F0E68C", c(red = 240, green = 230, blue = 140),
              "lavender", "#E6E6FA", c(red = 230, green = 230, blue = 250),
         "lavenderblush", "#FFF0F5", c(red = 255, green = 240, blue = 245),
             "lawngreen", "#7CFC00",   c(red = 124, green = 252, blue = 0),
          "lemonchiffon", "#FFFACD", c(red = 255, green = 250, blue = 205),
             "lightblue", "#ADD8E6", c(red = 173, green = 216, blue = 230),
            "lightcoral", "#F08080", c(red = 240, green = 128, blue = 128),
             "lighcyan", "#E0FFFF", c(red = 224, green = 255, blue = 255),
  "lightgoldenrodyellow", "#FAFAD2", c(red = 250, green = 250, blue = 210),
             "lightgray", "#D3D3D3", c(red = 211, green = 211, blue = 211),
             "lightgrey", "#D3D3D3", c(red = 211, green = 211, blue = 211),
            "lightgreen", "#90EE90", c(red = 144, green = 238, blue = 144),
             "lightpink", "#FFB6C1", c(red = 255, green = 182, blue = 193),
           "lightsalmon", "#FFA07A", c(red = 255, green = 160, blue = 122),
         "lightseagreen", "#20B2AA",  c(red = 32, green = 178, blue = 170),
          "lightskyblue", "#87CEFA", c(red = 135, green = 206, blue = 250),
        "lightslategray", "#778899", c(red = 119, green = 136, blue = 153),
        "lightslategrey", "#778899", c(red = 119, green = 136, blue = 153),
        "lightsteelblue", "#B0C4DE", c(red = 176, green = 196, blue = 222),
           "lightyellow", "#FFFFE0", c(red = 255, green = 255, blue = 224),
                  "lime", "#00FF00",     c(red = 0, green = 255, blue = 0),
             "limegreen", "#32CD32",   c(red = 50, green = 205, blue = 50),
                 "linen", "#FAF0E6", c(red = 250, green = 240, blue = 230),
               "magenta", "#FF00FF",   c(red = 255, green = 0, blue = 255),
                "maroon", "#800000",     c(red = 128, green = 0, blue = 0),
      "mediumaquamarine", "#66CDAA", c(red = 102, green = 205, blue = 170),
            "mediumblue", "#0000CD",     c(red = 0, green = 0, blue = 205),
          "mediumorchid", "#BA55D3",  c(red = 186, green = 85, blue = 211),
          "mediumpurple", "#9370DB", c(red = 147, green = 112, blue = 219),
        "mediumseagreen", "#3CB371",  c(red = 60, green = 179, blue = 113),
       "mediumslateblue", "#7B68EE", c(red = 123, green = 104, blue = 238),
     "mediumspringgreen", "#00FA9A",   c(red = 0, green = 250, blue = 154),
       "mediumturquoise", "#48D1CC",  c(red = 72, green = 209, blue = 204),
       "mediumvioletred", "#C71585",  c(red = 199, green = 21, blue = 133),
          "midnightblue", "#191970",   c(red = 25, green = 25, blue = 112),
             "mintcream", "#F5FFFA", c(red = 245, green = 255, blue = 250),
             "mistyrose", "#FFE4E1", c(red = 255, green = 228, blue = 225),
              "moccasin", "#FFE4B5", c(red = 255, green = 228, blue = 181),
           "navajowhite", "#FFDEAD", c(red = 255, green = 222, blue = 173),
                  "navy", "#000080",     c(red = 0, green = 0, blue = 128),
               "oldlace", "#FDF5E6", c(red = 253, green = 245, blue = 230),
                 "olive", "#808000",   c(red = 128, green = 128, blue = 0),
             "olivedrab", "#6B8E23",  c(red = 107, green = 142, blue = 35),
                "orange", "#FFA500",   c(red = 255, green = 165, blue = 0),
             "orangered", "#FF4500",    c(red = 255, green = 69, blue = 0),
                "orchid", "#DA70D6", c(red = 218, green = 112, blue = 214),
         "palegoldenrod", "#EEE8AA", c(red = 238, green = 232, blue = 170),
             "palegreen", "#98FB98", c(red = 152, green = 251, blue = 152),
         "palerturquoise", "#AFEEEE", c(red = 175, green = 238, blue = 238),
         "palevioletred", "#DB7093", c(red = 219, green = 112, blue = 147),
            "papayawhip", "#FFEFD5", c(red = 255, green = 239, blue = 213),
             "peachpuff", "#FFDAB9", c(red = 255, green = 218, blue = 185),
                  "peru", "#CD853F",  c(red = 205, green = 133, blue = 63),
                  "pink", "#FFC0CB", c(red = 255, green = 192, blue = 203),
                  "plum", "#DDA0DD", c(red = 221, green = 160, blue = 221),
            "powderblue", "#B0E0E6", c(red = 176, green = 224, blue = 230),
                "purple", "#800080",   c(red = 128, green = 0, blue = 128),
         "rebeccapurple", "#663399",  c(red = 102, green = 51, blue = 153),
                   "red", "#FF0000",     c(red = 255, green = 0, blue = 0),
             "rosybrown", "#BC8F8F", c(red = 188, green = 143, blue = 143),
             "royalblue", "#4169E1",  c(red = 65, green = 105, blue = 225),
           "saddlebrown", "#8B4513",   c(red = 139, green = 69, blue = 19),
                "salmon", "#FA8072", c(red = 250, green = 128, blue = 114),
            "sandybrown", "#F4A460",  c(red = 244, green = 164, blue = 96),
              "seagreen", "#2E8B57",   c(red = 46, green = 139, blue = 87),
              "seashell", "#FFF5EE", c(red = 255, green = 245, blue = 238),
                "sienna", "#A0522D",   c(red = 160, green = 82, blue = 45),
                "silver", "#C0C0C0", c(red = 192, green = 192, blue = 192),
               "skyblue", "#87CEEB", c(red = 135, green = 206, blue = 235),
             "slateblue", "#6A5ACD",  c(red = 106, green = 90, blue = 205),
             "slategray", "#708090", c(red = 112, green = 128, blue = 144),
             "slategrey", "#708090", c(red = 112, green = 128, blue = 144),
                  "snow", "#FFFAFA", c(red = 255, green = 250, blue = 250),
           "springgreen", "#00FF7F",   c(red = 0, green = 255, blue = 127),
             "steelblue", "#4682B4",  c(red = 70, green = 130, blue = 180),
                   "tan", "#D2B48C", c(red = 210, green = 180, blue = 140),
                  "teal", "#008080",   c(red = 0, green = 128, blue = 128),
               "thistle", "#D8BFD8", c(red = 216, green = 191, blue = 216),
                "tomato", "#FF6347",   c(red = 255, green = 99, blue = 71),
             "turquoise", "#40E0D0",  c(red = 64, green = 224, blue = 208),
                "violet", "#EE82EE", c(red = 238, green = 130, blue = 238),
                 "wheat", "#F5DEB3", c(red = 245, green = 222, blue = 179),
                 "white", "#FFFFFF", c(red = 255, green = 255, blue = 255),
            "whitesmoke", "#F5F5F5", c(red = 245, green = 245, blue = 245),
                "yellow", "#FFFF00",   c(red = 255, green = 255, blue = 0),
           "yellowgreen", "#9ACD32",  c(red = 154, green = 205, blue = 50)
  )

